File: RAS/RORSHTerminal/Crypto.cs
using System;
using System.Security.Cryptography;
using System.Text;

namespace RORSHTerminal
{
    /// <summary>
    /// Cryptographic utilities for RORSH Admin Shell
    /// Provides AES-256-GCM encryption, key derivation, and ECDH key exchange
    /// </summary>
    public static class Crypto
    {
        private const int KeySize = 32; // 256 bits
        private const int IvSize = 16;  // 128 bits
        private const int TagSize = 16; // 128 bits
        private const int SaltSize = 32;
        private const int Iterations = 100000;

        /// <summary>
        /// Generate a new ECDH key pair using NIST P-256 curve
        /// </summary>
        public static (string privateKey, string publicKey) GenerateKeyPair()
        {
            using (var ecdh = ECDiffieHellman.Create(ECCurve.NamedCurves.nistP256))
            {
                var privateKey = Convert.ToHexString(ecdh.ExportECPrivateKey());
                var publicKey = Convert.ToHexString(ecdh.ExportSubjectPublicKeyInfo());
                return (privateKey, publicKey);
            }
        }

        /// <summary>
        /// Derive AES-256 key from shared secret using PBKDF2
        /// </summary>
        public static string DeriveKey(string sharedSecret, byte[] salt = null)
        {
            if (salt == null)
            {
                salt = new byte[SaltSize];
                using (var rng = RandomNumberGenerator.Create())
                {
                    rng.GetBytes(salt);
                }
            }

            using (var pbkdf2 = new Rfc2898DeriveBytes(
                Encoding.UTF8.GetBytes(sharedSecret),
                salt,
                Iterations,
                HashAlgorithmName.SHA256))
            {
                var key = pbkdf2.GetBytes(KeySize);
                return Convert.ToHexString(key);
            }
        }

        /// <summary>
        /// Encrypt plaintext using AES-256-GCM
        /// Format: iv:tag:ciphertext (hex encoded)
        /// </summary>
        public static string Encrypt(string plaintext, string keyHex)
        {
            try
            {
                var key = Convert.FromHexString(keyHex);
                var iv = new byte[IvSize];
                using (var rng = RandomNumberGenerator.Create())
                {
                    rng.GetBytes(iv);
                }

                var plaintextBytes = Encoding.UTF8.GetBytes(plaintext);
                var ciphertext = new byte[plaintextBytes.Length];
                var tag = new byte[TagSize];

                using (var aes = new AesGcm(key, TagSize))
                {
                    aes.Encrypt(iv, plaintextBytes, ciphertext, tag);
                }

                return Convert.ToHexString(iv) + ":" + Convert.ToHexString(tag) + ":" + Convert.ToHexString(ciphertext);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[Crypto] Encryption error: {ex.Message}");
                return null;
            }
        }

        /// <summary>
        /// Decrypt ciphertext using AES-256-GCM
        /// Input: iv:tag:ciphertext (hex encoded)
        /// </summary>
        public static string Decrypt(string encryptedData, string keyHex)
        {
            try
            {
                var key = Convert.FromHexString(keyHex);
                var parts = encryptedData.Split(':');

                if (parts.Length != 3)
                {
                    throw new Exception("Invalid encrypted data format");
                }

                var iv = Convert.FromHexString(parts[0]);
                var tag = Convert.FromHexString(parts[1]);
                var ciphertext = Convert.FromHexString(parts[2]);
                var plaintext = new byte[ciphertext.Length];

                using (var aes = new AesGcm(key, TagSize))
                {
                    aes.Decrypt(iv, ciphertext, tag, plaintext);
                }

                return Encoding.UTF8.GetString(plaintext);
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[Crypto] Decryption error: {ex.Message}");
                return null;
            }
        }

        /// <summary>
        /// Compute SHA-256 hash
        /// </summary>
        public static string Sha256(string input)
        {
            using (var sha = SHA256.Create())
            {
                var hash = sha.ComputeHash(Encoding.UTF8.GetBytes(input));
                return Convert.ToHexString(hash);
            }
        }
    }
}

File: SecureCom.js/crypto.js
/**
 * RORSH Cryptographic Utilities
 * Provides AES-256-GCM encryption, key derivation, and RorshKey generation
 */

const crypto = require('crypto');

const ALGORITHM = 'aes-256-gcm';
const IV_LENGTH = 16;
const TAG_LENGTH = 16;
const SALT_LENGTH = 32;
const KEY_LENGTH = 32;
const ITERATIONS = 100000;

/**
 * Generate a random 10-digit RorshKey
 * Format: 10 numeric digits
 */
function generateRorshKey() {
    let key = '';
    for (let i = 0; i < 10; i++) {
        key += Math.floor(Math.random() * 10);
    }
    return key;
}

/**
 * Derive AES-256 key from a shared secret using PBKDF2
 */
function deriveKey(sharedSecret, salt = null) {
    if (!salt) {
        salt = crypto.randomBytes(SALT_LENGTH);
    } else if (typeof salt === 'string') {
        salt = Buffer.from(salt, 'hex');
    }

    const key = crypto.pbkdf2Sync(sharedSecret, salt, ITERATIONS, KEY_LENGTH, 'sha256');
    return {
        key: key.toString('hex'),
        salt: salt.toString('hex')
    };
}

/**
 * Encrypt payload using AES-256-GCM
 * Returns: iv + authTag + ciphertext (all hex encoded)
 */
function encryptPayload(plaintext, keyData) {
    try {
        const key = typeof keyData === 'string' ? Buffer.from(keyData, 'hex') : keyData.key ? Buffer.from(keyData.key, 'hex') : keyData;
        const iv = crypto.randomBytes(IV_LENGTH);
        const cipher = crypto.createCipheriv(ALGORITHM, key, iv);

        let encrypted = cipher.update(plaintext, 'utf8', 'hex');
        encrypted += cipher.final('hex');

        const tag = cipher.getAuthTag();

        // Format: iv:tag:ciphertext
        return iv.toString('hex') + ':' + tag.toString('hex') + ':' + encrypted;
    } catch (err) {
        console.error('Encryption error:', err.message);
        return null;
    }
}

/**
 * Decrypt payload using AES-256-GCM
 * Input: iv + authTag + ciphertext (hex encoded, colon separated)
 */
function decryptPayload(encryptedData, keyData) {
    try {
        const key = typeof keyData === 'string' ? Buffer.from(keyData, 'hex') : keyData.key ? Buffer.from(keyData.key, 'hex') : keyData;
        const parts = encryptedData.split(':');

        if (parts.length !== 3) {
            throw new Error('Invalid encrypted data format');
        }

        const iv = Buffer.from(parts[0], 'hex');
        const tag = Buffer.from(parts[1], 'hex');
        const ciphertext = parts[2];

        const decipher = crypto.createDecipheriv(ALGORITHM, key, iv);
        decipher.setAuthTag(tag);

        let decrypted = decipher.update(ciphertext, 'hex', 'utf8');
        decrypted += decipher.final('utf8');

        return decrypted;
    } catch (err) {
        console.error('Decryption error:', err.message);
        return null;
    }
}

/**
 * Generate ECDH key pair for session key exchange
 */
function generateKeyPair() {
    const ecdh = crypto.createECDH('secp256k1');
    ecdh.generateKeys();
    return {
        privateKey: ecdh.getPrivateKey('hex'),
        publicKey: ecdh.getPublicKey('hex')
    };
}

/**
 * Compute shared secret from private key and remote public key
 */
function computeShared(privateKeyHex, publicKeyHex) {
    const ecdh = crypto.createECDH('secp256k1');
    ecdh.setPrivateKey(privateKeyHex, 'hex');
    return ecdh.computeSecret(publicKeyHex, 'hex');
}

/**
 * Hash data using SHA-256
 */
function sha256(data) {
    return crypto.createHash('sha256').update(data).digest('hex');
}

module.exports = {
    generateRorshKey,
    deriveKey,
    encryptPayload,
    decryptPayload,
    generateKeyPair,
    computeShared,
    sha256
};
